# BIOF_501-Nextflow-Pipeline
## **By Steven Chen**
--- 
## Repository Contents
--- 
**Directories**
- `raw_data`: contains all the raw `idat` files and the sample sheet used as input for the workflow \
- `images`: contains the `.svg` file of the workflow summary image in this README.md file \
- `bins`: contains all the `.R` scripts that run the analysis \
- `results`: this results will be created by the workflow as the Nextflow pipeline runs, updating after each process step \
- `work`: this is generated by Nextflow and stores any intermediate files throughout the pipeline \

**Files**
- `methylationMain.nf`: this is the Nextflow file that contains all the processes and the workflow call itself \
- `nextflow.config`: this is the configuration file that sets some parameters for Nextflow, such as specifying the Docker container \
- `README.md`: this is the markdown file that provides usage and summary information regarding this repository \

## Pipeline Overview
--- 
This pipeline takes raw methylation .IDAT files from Illumina EPIC/EPICv2 array data using a comprehensive Nextflow workflow. Preprocessing, differential methylation probe (DMP) identification, and gene ontology (GO) enrichment analysis steps are performed using **R** packages to create an end-to-end analysis pipeline. 

## Background
---
Analyzing the human epigenome can reveal key factors and drivers of human health and disease. One integral component of epigenetic analysis is the investigation of methylation data. DNA methylation is the addition of methyl groups to nucleotides (most often cysteine), creating areas of methylation known as CpG sites. Gene methylation can lead to downstream regulatory effects on gene expression. Increased methylation (hypermethylation) has been associated with many diseases

Nicolaides-Baraister syndrome (NCBRS) is an epigenetic disease with various physical manifestations. It is caused by mutations in the SMARCA2 gene, a core component in epigenetic regulation [2]. especially important for methylation analysis =

An automated workflow/pipeline such as this one is necessary for the development of the field. Given the prevalence of bioinformatics packages across various languages, there are a plethora of different ways that raw methylation data could be analyzed. However, many individuals may not be well-versed or trained in bioinformatics, highlighting a need for an easily accessible, reproducible pipeline such as this. Moreover, reproducibility is a huge issue in scientific research, leading to many false and inflated results being published [1]. Thus, new pipelines such as this one must be created to ensure reproducibility and accessibility within the bioinformatics community, promoting best-practice for future research and fostering collaboration between members across scientific disciplines.

## Usage
--- 
### Dependencies
--- 
- Nextflow (version 24.10.1 or higher) - [Click here to install](https://www.nextflow.io/docs/latest/install.html)
- Docker (version 27.3.1 or higher) - [Click here to install](https://docs.docker.com/desktop/?_gl=1*1m4ai7c*_gcl_au*MTU4MTE4OTI5MS4xNzY0NDEyMjg1*_ga*MTcwMTkxNTU0NC4xNzY0MzI2Njk2*_ga_XJWPQMJYHQ*czE3NjQ0MTIyODUkbzIkZzEkdDE3NjQ0MTIzMDAkajQ1JGwwJGgw)
--- 
### Running the Workflow
1. Clone the repository
```bash
git clone https://github.com/anika-n-01/BIOF501_Project.git
```
2. Pull the raw methylation data from the NCBI GEO database (into the `/inputs` folder)
```bash
cd BIOF_501-Nextflow-Pipeline/inputs
```
```bash
curl -O https://ftp.ncbi.nlm.nih.gov/geo/series/GSE125nnn/GSE125367/suppl/GSE125367_RAW.tar
```
or 
```bash
wget https://ftp.ncbi.nlm.nih.gov/geo/series/GSE125nnn/GSE125367/suppl/GSE125367_RAW.tar
```

3. Untar and unzip the data
```bash
tar -xvf GSE125367_RAW.tar
gunzip *.gz
```

4. Run the nextflow pipeline from the `BIOF_501-Nextflow-Pipeline` directory
```bash
nextflow run methylationMain.nf
```

## Workflow Summary
---
1. **Preprocessing** using `minfi`\
a. *Filtering reads* - removing any poor quality reads (p < 0.05) \
b. *Normalization* - using quantile normalization \
c. *Failed probe removal* - some probes may have failed during sequencing, detect and remove these probes \
d. *SNP probe removals* - remove any probes with single nucleotide polymorphisms (SNPs) at the respective CpG site \
e. *Generate beta values* - generate methylation beta-values, which quantify the degree of hyper-methylation (>0) or hypo-methylation (<0) \

2. **Generate DMPs** \
a. *Create a design matrix* - design matrix created to compare between NCBRS and control, as well as covariates (e.g. age, sex) \
b. *Fit a linear model* - fit a linear model using the `limma` package \
c. *Generate DMPs* - use the model to generate a list of differentially methylated probes (DMPs) \

3. **GO analysis** \
a. *CpG annotation* - annotate the CpGs using the EPIC array annotation package\
b. *Filter significant DMPs* - filter for the most significant DMPs associated with NCBRS \
c. *GO enrichment analysis* - using the `missMethyl` package to perform GO enrichment \
d. *Generate plot* - using the `ggplot2` package to generate a plot of the top 10 GO terms associated with NCBRS in this data \

![Workflow Illustration](images/BIOF501Nextflow-2.svg)


## Expected Output
---

### Top 10 GO terms expected .csv file output 
![Here is the .csv output](images/Top10goterms.png)

### Top 10 GO terms plot output
![Here is the expected pdf output](images/GOplot.png)


## References
---
[1] Ioannidis, J. P. A. (2022). Correction: Why Most Published Research Findings Are False. PLoS Med, 19(8), e1004085. https://doi.org/10.1371/journal.pmed.1004085
[2] Van Houdt, J. K., Nowakowska, B. A., Sousa, S. B., van Schaik, B. D., Seuntjens, E., Avonce, N., Sifrim, A., Abdul-Rahman, O. A., van den Boogaard, M. J., Bottani, A., Castori, M., Cormier-Daire, V., Deardorff, M. A., Filges, I., Fryer, A., Fryns, J. P., Gana, S., Garavelli, L., Gillessen-Kaesbach, G.,…Vermeesch, J. R. (2012). Heterozygous missense mutations in SMARCA2 cause Nicolaides-Baraitser syndrome. Nat Genet, 44(4), 445–449, S441. https://doi.org/10.1038/ng.1105 

